File: algorithms/solutions/java/_0114_FlattenBinaryTreeToLinkedList.java
/*
 * @lc app=leetcode.cn id=114 lang=java
 *
 * [114] Flatten Binary Tree to Linked List
 *
 * https://leetcode.cn/problems/flatten-binary-tree-to-linked-list/description/
 *
 * algorithms
 * Medium (73.61%)
 * Likes:    1663
 * Dislikes: 0
 * Total Accepted:    461.4K
 * Total Submissions: 626.8K
 * Testcase Example:  '[1,2,5,3,4,null,6]'
 *
 * Given the root of a binary tree, flatten the tree into a "linked
 * list":
 * 
 * 
 * The "linked list" should use the same TreeNode class where the right child
 * pointer points to the next node in the list and the left child pointer is
 * always null.
 * The "linked list" should be in the same order as a pre-order traversal of
 * the binary tree.
 * 
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: root = [1,2,5,3,4,null,6]
 * Output: [1,null,2,null,3,null,4,null,5,null,6]
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: root = []
 * Output: []
 * 
 * 
 * Example 3:
 * 
 * 
 * Input: root = [0]
 * Output: [0]
 * 
 * 
 * 
 * Constraints:
 * 
 * 
 * The number of nodes in the tree is in the range [0, 2000].
 * -100 <= Node.val <= 100
 * 
 * 
 * 
 * Follow up: Can you flatten the tree in-place (with O(1) extra space)?
 */

public class _0114_FlattenBinaryTreeToLinkedList {
    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
// @lc code=start
    class Solution {
        public void flatten(TreeNode root) {
            if (root == null) return;
            flatten(root.left);
            flatten(root.right);
            // save previous right subtree
            TreeNode prevRight = root.right;
            // change root right pointer to root subtree
            root.right = root.left;
            root.left = null;
            // save previous right subtree to the rightest pointer
            while (root.right != null){
                root = root.right;
            }
            root.right = prevRight;
        }
    }
// @lc code=end
}

File: algorithms/solutions/python3/0114_flatten_binary_tree_to_linked_list.py
#
# @lc app=leetcode.cn id=114 lang=python3
#
# [114] Flatten Binary Tree to Linked List
#
# https://leetcode.cn/problems/flatten-binary-tree-to-linked-list/description/
#

# @lc code=start
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return None
        self.flatten(root.left)
        self.flatten(root.right)
        right = root.right # store the source right tree
        root.right = root.left
        root.left = None
        
        # find the rightest sub tree
        cur = root
        while cur.right:
            cur = cur.right
        cur.right = right
        
        
# @lc code=end

